libtenzir/src/tql2/set.cpp

```diff
@@ -296,13 +324,130 @@ auto assign(const ast::selector& left, series right, const table_slice& input,
     [&](const ast::meta& left) {
       return assign(left, right, input, dh);
     },
-    [&](const ast::simple_selector& left) {
+    [&](const ast::field_path& left) {
       auto result = std::vector<table_slice>{};
       result.push_back(assign(left, std::move(right), input, dh, position));
       return result;
     });
 }
 
+auto resolve_move_keyword(ast::assignment assignment)
+  -> std::pair<ast::assignment, std::vector<ast::field_path>> {
+  auto out = std::vector<ast::field_path>{};
+  auto recurse = [&](const auto& recurse, ast::expression& expr,
+                     std::vector<ast::field_path>& out) -> void {
+    match(
+      expr,
+      [&](ast::unary_expr& x) {
+        recurse(recurse, x.expr, out);
+        if (x.op.inner == ast::unary_op::move) {
+          if (auto field = ast::field_path::try_from(x.expr)) {
+            out.push_back(std::move(*field));
+            expr = std::move(x.expr);
+          }
+        }
+      },
+      [&](ast::record& x) {
+        for (auto& item : x.items) {
+          match(
+            item,
+            [&](ast::spread& x) {
+              recurse(recurse, x.expr, out);
+            },
+            [&](ast::record::field& x) {
+              recurse(recurse, x.expr, out);
+            });
+        }
+      },
+      [&](ast::list& x) {
+        for (auto& item : x.items) {
+          match(
+            item,
+            [&](ast::spread& x) {
+              recurse(recurse, x.expr, out);
+            },
+            [&](ast::expression& x) {
+              recurse(recurse, x, out);
+            });
+        }
+      },
+      [](ast::constant&) {}, [](ast::pipeline_expr&) {},
+      [](ast::root_field&) {}, [](ast::this_&) {}, [](ast::meta&) {},
+      [](ast::dollar_var&) {},
+      [&](ast::assignment& x) {
+        recurse(recurse, x.right, out);
+      },
+      [&](ast::unpack& x) {
+        recurse(recurse, x.expr, out);
+      },
+      [](ast::underscore&) {},
+      [&](ast::function_call& x) {
+        // TODO: The `move` and `where` functions abuse the expresions they
+        // take as arguments as a poor-mans lambda expression. We don't recurse
+        // further when we encounter them here, but that's at best a stopgap.
+        // Ideally, there'd be proper lambda support in the language itself.
+        if (x.fn.path.size() == 1) {
+          const auto& name = x.fn.path.front().name;
+          if (name == "move" or name == "where") {
+            return;
+          }
+        }
```

Comment:
Not sure how hard it'd be to add lambda expressions to the language including evaluator support. But we probably want to track this in a story referencing this comment.

Reply:
I expect closures to be a bit more involved. For example, what does a closure evaluate to? But I'd also like to have them at some point.

Reply:
Makes sense. I'll leave the comment in for future reference.